payroll/signals.py

```diff
@@ -11,143 +12,127 @@
-            if result \
-                    and result['success'] \
-                    and task['business_event'] == PayrollConfig.benefit_delete_event:
-                task_status = task['status']
-                if task_status == Task.Status.COMPLETED:
-                    benefit = BenefitConsumption.objects.get(id=task['entity_id'])
-                    delete_benefit(benefit, user)
-                if task_status == Task.Status.FAILED:
-                    benefit = BenefitConsumption.objects.get(id=task['entity_id'])
-                    benefit.status = BenefitConsumptionStatus.ACCEPTED
-                    benefit.save(username=user.username)
-        except Exception as exc:
-            logger.error("Error while executing on_task_complete_delete_benefit", exc_info=exc)
-
-    bind_service_signal(
-        'task_service.complete_task',
-        on_task_complete_accept_payroll,
-        bind_type=ServiceSignalBindType.AFTER
-    )
-
-    bind_service_signal(
-        'task_service.complete_task',
-        on_task_complete_payroll_reconcilation,
-        bind_type=ServiceSignalBindType.AFTER
-    )
-
-    bind_service_signal(
-        'task_service.complete_task',
-        on_task_complete_payroll_reject_approved_payroll,
-        bind_type=ServiceSignalBindType.AFTER
-    )
-
-    bind_service_signal(
-        'task_service.complete_task',
-        on_task_delete_payroll,
-        bind_type=ServiceSignalBindType.AFTER
-    )
-
-    bind_service_signal(
-        'task_service.complete_task',
-        on_task_delete_benefit,
-        bind_type=ServiceSignalBindType.AFTER
-    )
+        """
+        Handles the deletion of a benefit after task completion.
+        """
+        data = fetch_task_user_result(kwargs)
+        if not data or data["business_event"] != PayrollConfig.benefit_delete_event:
+            return
+
+        benefit = BenefitConsumption.objects.get(id=data['entity_id'])  # Get the benefit object
+        if data['status'] == Task.Status.COMPLETED:
+            StrategyOfPaymentInterface.remove_benefit_from_payroll(benefit)  # Remove benefit
+        elif data['status'] == Task.Status.FAILED:
+            benefit.status = BenefitConsumptionStatus.ACCEPTED  # Mark as accepted if failed
+            benefit.save(username=data['user'].username)
+
+    # Bind the signals to respective event handlers
+    bind_service_signal('task_service.complete_task', on_task_complete_accept_payroll, ServiceSignalBindType.AFTER)
+    bind_service_signal('task_service.complete_task', on_task_complete_payroll_reconciliation, ServiceSignalBindType.AFTER)
+    bind_service_signal('task_service.complete_task', on_task_complete_payroll_reject_approved_payroll, ServiceSignalBindType.AFTER)
+    bind_service_signal('task_service.complete_task', on_task_delete_payroll, ServiceSignalBindType.AFTER)
+    bind_service_signal('task_service.complete_task', on_task_delete_benefit, ServiceSignalBindType.AFTER)
```

Comment:
ideally this should be conditional in case someone want to make their own signal, those should not be registered